behave/helper/yapgdd/main.py

```diff
@@ -0,0 +1,230 @@
+import os
+import psycopg2
+import psycopg2.extras
+from urllib.parse import urlparse
+from tabulate import tabulate
+
+
+class Yapgdd:
+    def __init__(
+        self,
+        source_dsn,
+        target_dsn,
+        exclude_columns=[],
+        exclude_tables=[],
+        output_dir="/tmp",
+    ):
+        self.source_conn = self.get_connection(source_dsn)
+        self.target_conn = self.get_connection(target_dsn)
+        self.source_cursor = self.source_conn.cursor()
+        self.target_cursor = self.target_conn.cursor()
+        self.exclude_columns = set(exclude_columns)
+        self.exclude_tables = exclude_tables
+        self.output_dir = output_dir
+        os.makedirs(output_dir, exist_ok=True)
+
+        self.source_tables = self.get_tables(self.source_cursor)
+        self.target_tables = self.get_tables(self.target_cursor)
+        pass
+
+    def get_connection(self, dsn):
+        url = urlparse(dsn)
+        connection_details = {
+            "dbname": url.path.replace("/", ""),
+            "user": url.username,
+            "password": url.password,
+            "port": url.port,
+            "host": url.hostname,
+        }
+        return psycopg2.connect(**connection_details)
+
+    def get_dict_cursor(self, connection):
+        return connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+
+    def get_tables(self, cursor):
+        cursor.execute(
+            """SELECT table_name FROM information_schema.tables
+               WHERE table_schema = 'public'"""
+        )
+        return [table[0] for table in cursor.fetchall()]
+
+    def get_columns(self, cursor, table):
+        cursor.execute(f"Select * FROM {table} LIMIT 0")
+        return [desc[0] for desc in cursor.description]
+
+    def get_rows_count(self, cursor, table):
+        cursor.execute(f"Select count(*) FROM {table}")
+        return cursor.fetchone()[0]
+
+    def diff_row(self, table, source_row, target_row, row_number):
+        diff = {}
+        # Check source row matches target row
+        for key, value in source_row.items():
```

Comment:
just trying to think through what happens if something is in source with a value of None but not in target  - maybe covered by how it is called?

Reply:
Clarification: This is referring to when a column is removed from target

Reply:
This was already getting caught as a changed to column. I have added a test to capture this requirement